### ISSUE TITLE
assertions_on_constants is falsely reported on is_x86_feature_detected depending on target-cpu configuration

### ISSUE THREAD

**Issue by bdonlan** [rust-lang/rust-clippy#16434]
### Summary

When a `-Ctarget-cpu` option is passed to rustc, certain `is_x86_feature_detected!` invocations become compile-time constants. Clippy flags this code with `clippy::assertions_on_constants`, despite the constantness of the expression being dependent on the build configuration.

### Lint Name

assertions_on_constants

### Reproducer

I tried this code:

```rust
fn main() {
    assert!(std::arch::is_x86_feature_detected!("sse4.2"));
}
```

By default, no clippy assertion is issued (which is correct). If you then add the following to `.cargo/config.toml`:

```
[target.'cfg(target_arch = "x86_64")']
rustflags = ["-Ctarget-cpu=x86-64-v2"]
```

This now occurs
```
warning: this assertion has a constant value
 --> src/main.rs:2:5
  |
2 |     assert!(std::arch::is_x86_feature_detected!("sse4.2"));
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: consider moving this into a const block: `const { assert!(..) }`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.92.0/index.html#assertions_on_constants
  = note: `#[warn(clippy::assertions_on_constants)]` on by default
```

Further note that following the suggestion fails:
```
error[E0015]: cannot call non-const function `std_detect::detect::arch::x86::__is_feature_detected::sse4_2` in constants
 --> src/main.rs:2:21
  |
2 |     const { assert!(std::arch::is_x86_feature_detected!("sse4.2")) };
  |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: calls in constants are limited to constant functions, tuple structs and tuple variants
  = note: this error originates in the macro `std::arch::is_x86_feature_detected` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0015`.
error: could not compile `tmp-warning` (bin "tmp-warning") due to 1 previous error
```

### Version

```text
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```
